hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java

```diff
@@ -110,8 +127,110 @@ public SqlResult execute(@Nonnull SqlStatement statement) {
             return res;
         } catch (Exception e) {
             RuntimeException error = rethrow(e, connection);
-            res.onExecuteError(error);
-            throw error;
+            SqlResubmissionResult resubmissionResult = resubmitIfPossible(res, error);
+            if (resubmissionResult == null) {
+                res.onExecuteError(error);
+                throw error;
+            }
+            res.onResubmissionResponse(resubmissionResult);
+            return res;
+        }
+    }
+
+    SqlResubmissionResult resubmitIfPossible(SqlClientResult result, RuntimeException error) {
+        if (!shouldResubmit(error) || !shouldResubmit(result)) {
+            return null;
+        }
+
+        SqlResubmissionResult resubmissionResult = resubmitIfPossible0(result, error);
+        if (resubmissionResult.getSqlError() != null) {
+            SqlError sqlError = resubmissionResult.getSqlError();
+            throw new HazelcastSqlException(
+                    sqlError.getOriginatingMemberId(),
+                    sqlError.getCode(),
+                    sqlError.getMessage(),
+                    null,
+                    sqlError.getSuggestion()
+            );
+        }
+        return resubmissionResult;
+    }
+
+    @SuppressWarnings({"BusyWait", "checkstyle:cyclomaticcomplexity"})
+    private SqlResubmissionResult resubmitIfPossible0(SqlClientResult result, RuntimeException error) {
+        long resubmissionStartTime = System.nanoTime();
+        int invokeCount = 0;
+
+        SqlResubmissionResult resubmissionResult = null;
+        do {
+            ClientConnection connection = null;
+            try {
+                connection = getQueryConnection();
```

Comment:
What if this line keeps failing for 120 seconds? Then, we would return a `null` `resubmissionResult`, which is unexpected, as there are some places that use the result of this method without a null check